okta/resource_user.go

```diff
@@ -0,0 +1,672 @@
+}
+
+func assignAdminRolesToUser(u string, r []interface{}, c *okta.Client) error {
+	var valid bool
+
+	for _, role := range r {
+		valid = roleValidator(role.(string))
+		if valid {
+			roleStruct := okta.Role{Type: role.(string)}
+			_, _, err := c.User.AddRoleToUser(u, roleStruct, nil)
+
+			if err != nil {
+				return fmt.Errorf("[ERROR] Error Assigning Admin Roles to User: %v", err)
+			}
+		} else {
+			return fmt.Errorf("[ERROR] %v is not a valid Okta role", role)
+		}
+	}
+
+	return nil
+}
+
+// ValidateFunc in the schema package isn't supported for schema.TypeList yet so we'll use this helper function
+func roleValidator(r string) bool {
+	validRoles := []string{"SUPER_ADMIN", "ORG_ADMIN", "API_ACCESS_MANAGEMENT_ADMIN", "APP_ADMIN", "USER_ADMIN", "MOBILE_ADMIN", "READ_ONLY_ADMIN", "HELP_DESK_ADMIN"}
+
+	for _, v := range validRoles {
+		if v == r {
+			return true
+		}
+	}
+	return false
+}
+
+// need to remove from all current admin roles and reassign based on terraform configs when a change is detected
+func updateAdminRolesOnUser(u string, r []interface{}, c *okta.Client) error {
+	roles, _, err := c.User.ListAssignedRoles(u, nil)
+
+	if err != nil {
+		return fmt.Errorf("[ERROR] Error Updating Admin Roles On User: %v", err)
+	}
+
+	for _, role := range roles {
+		_, err := c.User.RemoveRoleFromUser(u, role.Id, nil)
+
+		if err != nil {
+			return fmt.Errorf("[ERROR] Error Updating Admin Roles On User: %v", err)
+		}
+	}
+
+	err = assignAdminRolesToUser(u, r, c)
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// camel cased strings from okta responses become underscore separated to match
+// the terraform configs for state file setting (ie. firstName from okta response becomes first_name)
+func camelCaseToUnderscore(s string) string {
+	a := []rune(s)
+
+	for i, r := range a {
+		if !unicode.IsLower(r) {
+			a = append(a, 0)
+			a[i] = unicode.ToLower(r)
+			copy(a[i+1:], a[i:])
+			a[i] = []rune("_")[0]
+		}
+	}
+
+	s = string(a)
+
+	return s
+}
+
+// regex lovingly lifted from: http://www.golangprograms.com/regular-expression-to-validate-email-address.html
+func matchEmailRegexp(val interface{}, key string) (warnings []string, errors []error) {
+	re := regexp.MustCompile("^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$")
+	if re.MatchString(val.(string)) == false {
+		errors = append(errors, fmt.Errorf("%s field not a valid email address", key))
+	}
+	return warnings, errors
+}
+
+// handle setting of user status based on what the current status is because okta
+// only allows transitions to certain statuses from other statuses - consult okta User API docs for more info
+// https://developer.okta.com/docs/api/resources/users#lifecycle-operations
+func updateUserStatus(u string, d string, c *okta.Client) error {
+	user, _, err := c.User.GetUser(u, nil)
+
+	if err != nil {
+		return err
+	}
+
+	switch d {
+	case "SUSPENDED":
```

Comment:
I guess you could save a little code here, drop a `var userErr error` in the upper scope, in your user calls in each case, use `userErr`, then check if it is nil below the case statement since they all do a single operation that could result in an error. Very minor but saves a few if statements!

Reply:
I dig it; going to change that up too - thanks for the feedback!